### PR TITLE
Dolfyn general updates

### PR DESCRIPTION
**Pull request by jmcvey3**
Fixes for:
- ADV skipped timesteps
- ADV attributes
- proper setting of max gap in ADV cleaning function
- error in integral length scale calculation
- error in despiking function
- move Reynolds stress and cross spectral density calculations to ADV folder
- prevent inaccurate TKE calculation from ADCP velocity input

I was updating this branch from master, so there may be repeated commits in this list, though that shouldn't matter after resolving conflicts and merging.

### CODE REVIEW

**Comment by jmcvey3** [comment]
Apparently I don't have "write access" to fix these conflicts

**Comment by ssolson** [comment]
> Apparently I don't have "write access" to fix these conflicts

I think that should only occur if you are trying to push changes to the `Develop` branch. You should be able to resolve these conflicts on your local machine via 
`git pull origin Develop`
***Fix the conflicts***
`git merge`
`git push jmcvey dolfyn`

**Comment by jmcvey3** [comment]
@ssolson Two things: currently netcdf4 is limited by the environment.yml file to <1.5.8, and I just discovered that if installing with conda, (`conda install -c conda-forge netcdf4`), version 1.5.7 is the latest conda allows.

**Comment by jmcvey3** [comment]
Having issues with this in tsdat, and the fundamental problem seems to be related to https://github.com/pydata/xarray/discussions/5709

**Comment by jmcvey3** [comment]
@ssolson Tests are failing because xarray is searching for a package called bottleneck https://pypi.org/project/Bottleneck/ https://docs.xarray.dev/en/v0.8.0/installing.html. Can we add this package to the requirements list?

**Comment by ssolson** [comment]
Looks like it is an optional xarray package used to speed up working with NaNs which seems like it would be useful for DOLfYN. I checked out the repo and they seem active and have working tests for all current python versions. I think it should be good to add.

**Comment by ssolson** [comment]
Hey, James how goes this PR? Do you need any help?

**Comment by jmcvey3** [comment]
Hi Sterling, this update should be good to go, once I'm sure the tests pass online.
Related - is there a different process for updating documentation? I can put a pull request in for that as well.

**Comment by jmcvey3** [comment]
@ssolson I do need help with fixing the "pip-windows-latest" tests. Should be the last thing.

**Comment by ssolson** [comment]
@jmcvey3 I am working on a fix.

File: mhkit/dolfyn/tools/fft.py
Comment (akeeste) [inline mhkit/dolfyn/tools/fft.py:80]: Was this function moved/renamed, or removed entirely?

File: mhkit/dolfyn/tools/fft.py
Comment (akeeste) [inline mhkit/dolfyn/tools/fft.py:148]: Was this function moved/renamed, or removed entirely?

File: mhkit/dolfyn/velocity.py
Comment (akeeste) [inline mhkit/dolfyn/velocity.py:592]: Was this function moved/renamed, or removed entirely?

File: mhkit/dolfyn/velocity.py
Comment (akeeste) [inline mhkit/dolfyn/velocity.py:672]: Was this function moved/renamed, or removed entirely?

File: mhkit/dolfyn/velocity.py
Comment (akeeste) [inline mhkit/dolfyn/velocity.py:812]: Was this function moved/renamed, or removed entirely?

File: mhkit/dolfyn/velocity.py
Comment (akeeste) [inline mhkit/dolfyn/velocity.py:679]: Is there an alternative that won't work break in the ADCP 4-beam case?

File: mhkit/dolfyn/velocity.py
Comment (akeeste) [inline mhkit/dolfyn/velocity.py:941]: Was this function moved/renamed, or removed entirely?

File: mhkit/dolfyn/velocity.py
Comment (akeeste) [inline mhkit/dolfyn/velocity.py:1073]: Was this function moved/renamed, or removed entirely?

**Review by akeeste — COMMENTED** [review]
Hi @jmcvey3

I'm reviewing this for @ssolson to help out on his PR workload. As you say in your initial description, these are primarily refactoring changes. This PR does not add fundamentally new functionality, correct? I'm not familiar with ADVs or ADCPs so I can't comment much on the packages from that perspective.

The biggest thing I noted is that there were several functions removed from /tools/fft.py and tools/velocity.py. Some of these were added to adp/turbulence.py, but I'm not seeing all of them. Can you confirm where the functions removed from /tools/ went to? I commented on those I could not track. GitHub wasn't comparing text very well so I'm sure I missed a few things.

Adam

File: mhkit/dolfyn/binned.py
Comment (jmcvey3) [reply mhkit/dolfyn/binned.py:331]: The relevant functions have been updated in this pull request from what I just searched?

File: mhkit/dolfyn/velocity.py
Comment (jmcvey3) [reply mhkit/dolfyn/velocity.py:679]: Ah, it should error for ADCP's, and I should add an exception here

File: mhkit/dolfyn/velocity.py
Comment (jmcvey3) [reply mhkit/dolfyn/velocity.py:941]: I renamed to `reynolds_stress` to make it more descriptive and moved to the ADV/turbulence.py file because this algorithm (simply the covariance between velocity signals) is only accurate for point-measurements (ADV, shear probe, etc).

File: mhkit/dolfyn/velocity.py
Comment (jmcvey3) [reply mhkit/dolfyn/velocity.py:1073]: Moved this to ADV/turbulence.py for the same reason above

**Comment by jmcvey3** [comment]
Hi Adam,

Just left comments on the things you flagged. I removed a number of signal processing functions from velocity.py and tools/ because they're built-in to other packages like scipy (but weren't when dolfyn was originally written).

This pull request will add additional functionality to mhkit because only the basic processing functionality is available right now. I have a few more updates to add to this PR (or a new one) from working with Emily's Tanana River data and Levi's Cook Inlet data. I'm waiting on test data from Emily for those updates, and verification from Levi.

James

**Comment by akeeste** [comment]
@jmcvey3 Thanks for clarifying and helping track where those functions moved to. That resolves all of my comments. If this PR is otherwise ready to go and more large updates are coming, I think we should merge now and create a new PR for the next updates.

**Comment by ssolson** [comment]
@jmcvey3 could you merge the latest from `Develop` into this PR to see if it fixes your pip install issues?

**Comment by jmcvey3** [comment]
Great, sounds good to me Adam

**Review by ssolson — COMMENTED** [review]
@jmcvey3 I will merge this into the Develop branch.

For you next PR can you look have the update the docstrings to go from this

```
        """Calculate the cross-spectral density of velocity components.
        Parameters
        ----------
        veldat   : xarray.DataArray
          The raw 3D velocity data.

        Returns
        -------
        csd : xarray.DataArray (3, M, N_FFT)
          The first-dimension of the cross-spectrum is the three
          different cross-spectra: 'uv', 'uw', 'vw'.

        """
        x = 1+2
```
to this
```
        """
        Calculate the cross-spectral density of velocity components.
        
        Parameters
        ----------
        veldat   : xarray.DataArray
          The raw 3D velocity data.

        Returns
        -------
        csd : xarray.DataArray (3, M, N_FFT)
          The first-dimension of the cross-spectrum is the three
          different cross-spectra: 'uv', 'uw', 'vw'.
        """

        x = 1+2
```
Namely
1. Start docstring description on a new line
2. No newline at end of docstring
3. New line directly following the end of the docstring

**Review by ssolson — APPROVED** [review]

**Pull request merged**
